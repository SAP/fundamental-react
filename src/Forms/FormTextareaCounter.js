import classnames from 'classnames';
import CustomPropTypes from '../utils/CustomPropTypes/CustomPropTypes';
import PropTypes from 'prop-types';
import React from 'react';

const FormTextareaCounter = React.forwardRef(({ children, className, charCount, localizedText, showCounter, ...props }, ref) => {

    const formTextareaCounterClasses = classnames(
        'fd-textarea-counter', className
    );

    const getText = () => {
        let text = children || '';
        let keyword = localizedText.counterKeyword,
            keywordIdx = text.indexOf(keyword),
            count = charCount >= 0 ? charCount : 0;
        if (keywordIdx === -1) {
            text = localizedText.counterText;
            keywordIdx = text.indexOf(keyword);
        }
        if (keywordIdx >= 0) {
            text = keywordIdx + keyword.length < text.length ?
                text.substring(0, keywordIdx) + count + text.substring(keywordIdx + keyword.length) :
                text.substring(0, keywordIdx) + count;
        }
        return text;
    };

    return (showCounter ?
        <div {...props}
            className={formTextareaCounterClasses}
            ref={ref}>
            {getText()}
        </div> : null);
});

FormTextareaCounter.displayName = 'FormTextareaCounter';

FormTextareaCounter.propTypes = {
    charCount: PropTypes.number,
    children: PropTypes.node,
    className: PropTypes.string,
    localizedText: CustomPropTypes.i18n({
        counterKeyword: PropTypes.string,
        counterText: PropTypes.string
    }),
    showCounter: PropTypes.bool
};

FormTextareaCounter.defaultProps = {
    localizedText: {
        counterKeyword: '#counter#',
        counterText: '#counter# characters left'
    }
};

export default FormTextareaCounter;
