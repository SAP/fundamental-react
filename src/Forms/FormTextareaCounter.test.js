import FormTextareaCounter from './FormTextareaCounter';
import { mount } from 'enzyme';
import React from 'react';
import renderer from 'react-test-renderer';

describe('<FormTextAreaCounter />', () => {

    test('create form textareacounter', () => {
        let component = renderer.create(mount(<FormTextareaCounter
            charCount={150}
            showCounter />));
        let tree = component.toJSON();
        expect(tree).toMatchSnapshot();
    });

    describe('Prop spreading', () => {
        test('should allow props to be spread to the FormTextareaCounter component', () => {
            const element = mount(<FormTextareaCounter data-sample='Sample' showCounter />);
            expect(
                element.getDOMNode().attributes['data-sample'].value
            ).toBe('Sample');
        });
    });

    test('forwards the ref', () => {
        let ref;
        class Test extends React.Component {
            constructor(props) {
                super(props);
                ref = React.createRef();
            }
            render = () => <FormTextareaCounter ref={ref} showCounter />;
        }
        mount(<Test />);
        expect(ref.current.tagName).toEqual('DIV');
        expect(ref.current.className).toEqual('fd-textarea-counter');
    });

    describe('TextareaCounter tests default', () => {
        let setup = (props) => {
            return mount(<FormTextareaCounter {...props} />);
        };
        test('should add attribute when character count is passed to FormTextarea', () => {
            let element = setup({
                charCount: 150
            });
            expect(element.props().charCount).toBe(150);
        });
        test('should display localized text by default', () => {
            let element = setup({
                charCount: 1,
                localizedText: {
                    counterText: '#placeholder# chars left',
                    counterKeyword: '#placeholder#'
                },
                showCounter: true
            });
            expect(element.find('.fd-textarea-counter').text()).toEqual('1 chars left');
        });
    });
    describe('TextareaCounter tests custom text', () => {
        let setup = (props, text) => {
            return mount(<FormTextareaCounter {...props}>{text}</FormTextareaCounter>);
        };

        test('should display counter at the end', () => {
            let element = setup({
                charCount: 1,
                localizedText: {
                    counterKeyword: '#count#',
                    counterText: 'something'
                },
                showCounter: true
            }, 'word count #count#');
            expect(element.find('.fd-textarea-counter').text()).toEqual('word count 1');
        });

        test('should display only counter', () => {
            let element = setup({
                charCount: 1,
                showCounter: true
            }, '#counter#');
            expect(element.find('.fd-textarea-counter').text()).toEqual('1');
        });
        test('should use localizedText when keyword not found in custom text', () => {
            let element = setup({
                charCount: 1,
                localizedText: {
                    counterText: '#counter# characters left',
                    counterKeyword: '#counter#' },
                showCounter: true
            }, 'chars left #invalidkeyword');
            expect(element.find('.fd-textarea-counter').text()).toEqual('1 characters left');
        });
    });
});
