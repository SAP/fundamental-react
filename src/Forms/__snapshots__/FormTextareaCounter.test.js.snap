// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<FormTextAreaCounter /> create form textareacounter 1`] = `
<div
  className="fd-textarea-counter"
>
  150 characters left
</div>
`;
